hw_11/HW_1_Sotnikov.py

```diff
@@ -0,0 +1,368 @@
+            attr = attrs[0]
+            value = float(attrs[1])
+            return type(self)(self.operands[attr](self, value))
+
+number = StrangeFloat(3.5)
+
+number.add_1
+
+number.subtract_20
+
+number.multiply_5
+
+number.divide_25
+
+number.add_1.add_2.multiply_6.divide_8.subtract_9
+
+getattr(number, "add_-2.5")   # Используем getattr, так как не можем написать number.add_-2.5 - это SyntaxError
+
+
+# 4 Dunders vs normals
+
+## Before
+
+import numpy as np
+
+
+matrix = []
+for idx in range(0, 100, 10):
+    matrix += [list(range(idx, idx + 10))]
+
+selected_columns_indices = list(filter(lambda x: x in range(1, 5, 2), range(len(matrix))))
+selected_columns = map(lambda x: [x[col] for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr[:, 1] % 3 == 0
+new_arr = arr[mask]
+
+product = new_arr @ new_arr.T
+
+if (product[0] < 1000).all() and (product[2] > 1000).any():
+    print(product.mean())
+
+## After
+
+np = __import__("numpy")
+
+
+matrix = []
+for idx in range(0, 100, 10):
+    matrix.__iadd__([list(range(idx, idx.__add__(10)))])
+
+selected_columns_indices = list(filter(lambda x: x in range(1, 5, 2), range((matrix).__len__())))
+selected_columns = map(lambda x: [x.__getitem__(col) for col in selected_columns_indices], matrix)
+
+arr = np.array(list(selected_columns))
+
+mask = arr[:, 1].__mod__(3).__eq__(0)
+new_arr = arr.__getitem__(mask)
+
+product = new_arr.__matmul__(new_arr.T)
+
+if (product.__getitem__(0).__lt__(1000)).all().__and__((product.__getitem__(2).__gt__(1000)).any()):
+    print(product.mean())
+
+
+# 5 Biopython na minimalkah
+
+## Creating abstract class
+from abc import ABC, abstractmethod
+
+
+class BiologicalSequence(ABC):
+    @abstractmethod
+    def __len__(self):
+        pass
+
+    @abstractmethod
+    def __getitem__(self, slc):
+        pass
+
+    @abstractmethod
+    def __str__(self):
+        pass
+
+    @abstractmethod
+    def __repr__(self):
+        pass
+
+    @abstractmethod
+    def correctness(self):
+        pass
+
+class IncorrectSequenceAlphabet(ValueError):
+    pass
+
+
+## Creating nucleotide class
+
+class NucleicAcidSequence(BiologicalSequence):
+
+    def __len__(self):
+        return len(self.seq)
+
+    def __getitem__(self, slc):
+        return self.seq[slc]
+
+    def __str__(self):
+        return str(self.seq)
+
+    def __repr__(self):
+        return str(f"Sequence {self.seq}")
+
+    def __upper__(self):
+        return self.seq.upper()
+
+    def __lower__(self):
+        return self.seq.lower()
+
+    def correctness(self):
+        if any(monomer not in self.list for monomer in self.seq.upper()):
+            raise IncorrectSequenceAlphabet("Your sequence contains illegal symbols.")
+        else:
+            return True
+
+    def complement(self):
+        if self.correctness():
+            result = ''.join([self.dictionary[nucleotide] for nucleotide in self.seq])
+        return result
```

Comment:
При комплементировании мы ожидаем получить данные такого же типа, а не строку
```suggestion
            result = ''.join([self.dictionary[nucleotide] for nucleotide in self.seq])
        return type(self)(result)
```